sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/UnionAwareOptimizerRule.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.execution.{SparkPlan, UnionExec}
+
+trait UnionAwareOptimizerRule {
```

Comment:
> Can we make this rule more general? Ideally we need to split the query plan of the given query stage into several groups. The shuffles within one group must have the same number of partitions.
> 
> This rule is overly simplified right now and assumes the entire query stage is one group. It's definitely wrong with Union and we should fix it. To be conservative we can try to put everything in one group except for Union. We should make the code extensible so that we can add more "breaking points" that can split groups in the future

@cloud-fan agree with it, try to add a new trait to make it clear. I believe other query stage optimizer rule e.g., `OptimizeSkewedJoin` also need this. Do you have any thought ?

Reply:
We can always add abstraction later when we need to reuse the code in other places. For now let's focus on the coalesce shuffle partitions rule first.

Reply:
Back to code, I'm not very sure this approach is good enough for the more general requirement. At least, the idea here is 
1. Optimize using the Union's children that we can treat them as the atomic plan which can be optimized
2. Optimize using the whole plan if step 1 is not satisfied

Reply:
We can simplify the logic of this rule and only ensure that the children in the Union can be optimized. We don’t need to consider whether the child of the Union also contains the union, because when we call rule(child), we will consider the children of the Union recursively.

Reply:
Thank you for the input. Actual the rule(child) does not consider the children of the Union, as it just try to optimize the plan.

For the nested Union. Here are two case, the first is ok if we skip the check but the second can be optimized through every Union that will cause the repetitiion.
```
Union
  HashAggregate
    ShuffleQueryStage
  Union
    HashAggregate
      ShuffleQueryStage
    FileScan
```

```
Union
  HashAggregate
    ShuffleQueryStage
  Union
    HashAggregate
      ShuffleQueryStage
    HashAggregate
      ShuffleQueryStage
```